### PR TITLE
Calculate flow rate 118362289

### PR DESCRIPTION
**Pull request by brennanneoh**
@rpbaltazar Have a look 😄

### CODE REVIEW

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 1]
```diff
@@ -44,7 +44,7 @@ describe 'orifice-calculator-viewmodel-spec', ->
       expect(viewModel.pipeID().length).toEqual 6
 
     it 'should have default value of "1.939\'\' XS, Sch 80, Sch 80S"', ->
-      expect(viewModel.selectedPipeID()).toEqual OPL.OrificeCalculator.Config.Dictionaries.PipeID.oneNineInch
+      expect(viewModel.selectedPipeID()).toEqual 1.939
```
Comment (rpbaltazar) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Please keep using the dictionary variables instead

File: src/js/ko-components/config/dictionaries.coffee [hunk 2]
```diff
@@ -3,16 +3,19 @@
...
-  oneNineInch: "1.939'' XS, Sch 80, Sch 80S" # Default
-  twoOneInch: "2.157'' Sch 10, Sch 10S"
-  twoTwoInch: "2.245'' Sch 5, Sch 5S"
-  oneSixInch: "1.687'' Sch 160"
-  oneHalfInch: "1.503'' XXS"
-}
-@OPL.OrificeCalculator.Config.Dictionaries.OperatingPressureRead = { gauge: 'Gauge', absolute: 'Absolute' }
-@OPL.OrificeCalculator.Config.Dictionaries.OperatingPressureUnits = {
+@OPL.OrificeCalculator.Config.Dictionaries.PipeID =
+  2.067: "2.067'' Sch 40, STD, Sch 40S"
```
Comment (rpbaltazar) [inline src/js/ko-components/config/dictionaries.coffee]: You should not use . on variable naming.
. is used as namespacing and has a very specific meaning. please find another way of dealing with this

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 3]
```diff
@@ -62,3 +66,22 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
 
     @compressibilityCorrection = ko.observableArray _.values OPL.OrificeCalculator.Config.Dictionaries.CompressibilityCorrection
     @chosenCompressibilityCorrection  = ko.observable OPL.OrificeCalculator.Config.Dictionaries.CompressibilityCorrection.none
+    @compressibilityCorrectionValue = ko.observable(1)
+    @betaRatio = ko.computed =>
+      @orificeBoreDiameter() / @selectedPipeID()
+
+    @velocityOfApproach = ko.computed =>
+      Math.ceil((1 / Math.sqrt 1 - @betaRatio() ** 4) * 10 ** 2) / 10 ** 2
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Where did you get the velocity of approach equation from?

Please add parenthesis to math.sqrt to emphasize which params should be in the SQRT.
`Math.ceil((1 / Math.sqrt(1 - @betaRatio() ** 4)) * 10 ** 2) / 10 ** 2`

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 4]
```diff
@@ -62,3 +66,22 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
+
+    @flowRate = ko.computed =>
+      coeffDischarge = 0.6
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: These should be constants defined in the beginning of the viewModel. makes it more clear for whoever comes next

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 5]
```diff
@@ -62,3 +66,22 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
+      expansionFactor = 1
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Same thing applies

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 6]
```diff
@@ -62,3 +66,22 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
+      baseTemperature = 519.67 # Tb in Rankine, assumed 60F
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: same thing applies

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 7]
```diff
@@ -62,3 +66,22 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
+      # TODO: make observable return integer instead of string by default?
+      operatingTemperatureInRankine = Number(@operatingTemperature()) + 459.67
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: The 459.67 constant im pretty sure it has a name and could be moved to the constants as well.

All the other constants should be moved

File: src/js/ko-components/view-models/orifice-calculator.coffee (hunk 4 above)
Comment (brennanneoh) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: For reference: https://coderwall.com/p/1dckba/constants-in-coffeescript

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 8]
```diff
@@ -19,6 +17,12 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
         when 200 < @operatingPressure() <= 400 then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
         when 401 <= @operatingPressure() then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
 
+    @pipeID = ko.observableArray _.map OPL.OrificeCalculator.Config.Dictionaries.PipeID, (pipeName, pipeDiameter) ->
+      name: pipeName
+      value: pipeDiameter
+
+    @selectedPipeID = ko.observable 1.939
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Why would we even have a dictionary if then we refer to the values magically in the code?

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 9]
```diff
@@ -19,6 +17,12 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
         when 200 < @operatingPressure() <= 400 then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
         when 401 <= @operatingPressure() then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
 
+    @pipeID = ko.observableArray _.map OPL.OrificeCalculator.Config.Dictionaries.PipeID, (pipeName, pipeDiameter) ->
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: And this mapping?

File: src/js/ko-components/config/dictionaries.coffee (hunk 2 above)
Comment (thekcsam) [inline src/js/ko-components/config/dictionaries.coffee]: Noted with it. Done with correction.

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee (hunk 1 above)
Comment (thekcsam) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Noted with it. Done with correction.

**Comment by thekcsam** [comment]
@rpbaltazar, we have done the correction. Please have a look. 😄

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 10]
```diff
@@ -117,3 +117,44 @@ describe 'orifice-calculator-viewmodel-spec', ->
...
+      viewModel.orificeBoreDiameter 0.776
+      viewModel.selectedPipeID OPL.OrificeCalculator.Config.Dictionaries.PipeID.oneNineInch.value
+      viewModel.operatingPressure 900
+      viewModel.compressibilityCorrectionValue 1
+      viewModel.differentialPressure 30
+      viewModel.baseSpecificGravity 1
+      viewModel.operatingTemperature 60
+      expect(viewModel.flowRate()).toEqual 341.328
+
+  describe 'Math.ceil10', ->
```
Comment (rpbaltazar) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Why are you testing that javascript works as supposed? Did you invent this method? No.
So no need to test

Ok, realized now that it is not part of the standard. Please use https://lodash.com/docs#ceil instead

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 11]
```diff
@@ -3,9 +3,17 @@
 @OPL.KoComponents.ViewModels ||= {}
 
 class OPL.KoComponents.ViewModels.OrificeCalculator
+  COEFFICIENT_OF_DISCHARGE = 0.6
+  EXPANSION_FACTOR         = 1
+  BASE_TEMPERATURE         = 519.67
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Lets just add some notes here so we don't forget which unit is stored in this constant, please
Same thing applies to base pressure and absolute zero

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 12]
```diff
@@ -62,3 +70,31 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
 
     @compressibilityCorrection = ko.observableArray _.values OPL.OrificeCalculator.Config.Dictionaries.CompressibilityCorrection
     @chosenCompressibilityCorrection  = ko.observable OPL.OrificeCalculator.Config.Dictionaries.CompressibilityCorrection.none
+    @compressibilityCorrectionValue = ko.observable(1)
+
+    Math.ceil10 ||= {}
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Please use https://lodash.com/docs#ceil instead

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 13]
```diff
@@ -62,3 +70,33 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
+    @velocityOfApproach = ko.computed =>
+      Math.ceil10(1 / Math.sqrt(1 - @betaRatio() ** 4), -2)
+
+    @flowRate = ko.computed =>
+      operatingTemperatureInRankine = +@operatingTemperature() + ABSOLUTE_ZERO
+
+      flowRate = 218.527 * COEFFICIENT_OF_DISCHARGE * EXPANSION_FACTOR * @velocityOfApproach() * @orificeBoreDiameter() ** 2 * BASE_TEMPERATURE/BASE_PRESSURE \
+                 * ((@operatingPressure() * BASE_COMPRESSIBILITY * @differentialPressure()) \
+                 / (@baseSpecificGravity() * @compressibilityCorrectionValue() * operatingTemperatureInRankine)) ** 0.5 / 60
+      Math.ceil10(flowRate, -3)
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: use _.ceil instead

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 14]
```diff
@@ -62,3 +70,33 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
+    @betaRatio = ko.computed =>
+      Math.ceil10(@orificeBoreDiameter() / @selectedPipeDiameter(), -2)
+
+    @velocityOfApproach = ko.computed =>
+      Math.ceil10(1 / Math.sqrt(1 - @betaRatio() ** 4), -2)
+
+    @flowRate = ko.computed =>
+      operatingTemperatureInRankine = +@operatingTemperature() + ABSOLUTE_ZERO
+
+      flowRate = 218.527 * COEFFICIENT_OF_DISCHARGE * EXPANSION_FACTOR * @velocityOfApproach() * @orificeBoreDiameter() ** 2 * BASE_TEMPERATURE/BASE_PRESSURE \
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: 218.527 please move it to a constant as well -> `UNIT_CONVERSION_FACTOR`

File: spec/specrunner.haml [hunk 15]
```diff
@@ -4,7 +4,7 @@
   %body
     %script{:src => "https://cdnjs.cloudflare.com/ajax/libs/knockout/3.3.0/knockout-min.js"}
     %script{:src => 'https://cdnjs.cloudflare.com/ajax/libs/knockout-validation/2.0.3/knockout.validation.min.js' }
-    %script{:src => "https://cdnjs.cloudflare.com/ajax/libs/lodash.js/3.9.0/lodash.min.js"}
+    %script{:src => "https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.12.0/lodash.min.js"}
```
Comment (brennanneoh) [inline spec/specrunner.haml]: _.ceil is not available in lodash 3.9.0.

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 16]
```diff
@@ -3,10 +3,24 @@
 @OPL.KoComponents.ViewModels ||= {}
 
 class OPL.KoComponents.ViewModels.OrificeCalculator
+  UNIT_CONVERSION_FACTOR = 218.527
+  COEFFICIENT_OF_DISCHARGE = 0.6
+  EXPANSION_FACTOR         = 1
+
+  # NOTE: The base temperature is in degree rankine
+  BASE_TEMPERATURE         = 519.67
+  BASE_PRESSURE            = 14.73
```
Comment (thekcsam) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Add note to this.

File: spec/js/spec-helper.coffee [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    'orifice-calculator-viewmodel':
+      deps: ['knockout.validation', 'orifice-calculator-config']
+    'orifice-calculator-viewmodel-spec':
+      deps: ['orifice-calculator-viewmodel']
+    'orifice-calculator-view-spec':
+      deps: ['jasmine-jquery']
+}
+
+require ['orifice-calculator-viewmodel-spec', 'orifice-calculator-view-spec'], () ->
+    window.onload()
```
Comment (rpbaltazar) [inline spec/js/spec-helper.coffee]: Why do we need this?

Comment (rpbaltazar) [inline spec/js/spec-helper.coffee]: Fix this indentation

Comment (thekcsam) [inline spec/js/spec-helper.coffee]: From [this article](https://www.airpair.com/jasmine/posts/javascriptintegrating-jasmine-with-requirejs-amd):

> we trigger window.onload() as Jasmine hooks into this event to initialise its engine

**Comment by thekcsam** [comment]
Merged by @brennanneoh and @samkahchiin

**Pull request closed without merging**